File: sources/vk_parser.py
"""
Парсер ВКонтакте — получение вакансий из стен VK-групп через API.

Использует официальное VK API v5.131:
  - utils.resolveScreenName — получение ID группы по короткому имени
  - wall.get — получение постов со стены

Все запросы асинхронные через aiohttp. Каждая группа обрабатывается
независимо, ошибки в одной не блокируют остальные.
"""

from __future__ import annotations

import hashlib
import os
import re
import sys
from datetime import datetime, timedelta, timezone
from typing import Any

import aiohttp
from loguru import logger

# ── Гарантируем доступ к config из корня проекта ──────────────
sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))

from config import (  # noqa: E402
    HOURS_LOOKBACK,
    VK_ACCESS_TOKEN,
    VK_API_VERSION,
    VK_GROUPS_TO_MONITOR,
)

# ── Константы VK API ─────────────────────────────────────────
_VK_API_BASE = "https://api.vk.com/method"

# Паттерн для извлечения URL
_URL_PATTERN = re.compile(r"https?://[^\s<>\"')\]]+", re.IGNORECASE)

# Паттерн для поиска зарплаты: «50000», «50 000», «50к», «50k», «50 тыс»
_SALARY_CONTEXT_PATTERN = re.compile(
    r"(?:зарплата|оплата|оклад|з/?п|зп|ставка|доход)"
    r".{0,50}"
    r"(\d[\d\s]*\d|\d+)"
    r"\s*(000|к|k|тыс\.?|руб\.?|₽)?",
    re.IGNORECASE,
)
_SALARY_RUB_PATTERN = re.compile(
    r"(\d[\d\s]*\d|\d+)\s*(000|к|k|тыс\.?)?\s*(?:₽|руб)",
    re.IGNORECASE,
)


def _extract_salary(text: str) -> tuple[int | None, int | None]:
    """
    Извлечь диапазон зарплаты из текста поста.

    Returns:
        Кортеж (salary_from, salary_to). Если не найдено — (None, None).
    """
    matches = _SALARY_CONTEXT_PATTERN.findall(text)
    if not matches:
        matches = _SALARY_RUB_PATTERN.findall(text)
    if not matches:
        return None, None

    values: list[int] = []
    for raw_number, suffix in matches[:4]:
        try:
            number = int(raw_number.replace(" ", "").replace("\u00a0", ""))
        except ValueError:
            continue

        suffix_lower = suffix.lower().rstrip(".")
        if suffix_lower in ("к", "k", "тыс"):
            number *= 1000

        if 5_000 <= number <= 1_000_000:
            values.append(number)

    if not values:
        return None, None
    if len(values) == 1:
        return values[0], None
    return min(values), max(values)


async def _vk_api_request(
    session: aiohttp.ClientSession,
    method: str,
    params: dict[str, Any],
) -> dict[str, Any] | None:
    """
    Выполнить запрос к VK API.

    Args:
        session: aiohttp-сессия
        method: имя метода VK API (например 'wall.get')
        params: параметры запроса

    Returns:
        Словарь с ответом или None при ошибке.
    """
    params.setdefault("access_token", VK_ACCESS_TOKEN)
    params.setdefault("v", VK_API_VERSION)

    url = f"{_VK_API_BASE}/{method}"

    try:
        async with session.get(url, params=params, timeout=aiohttp.ClientTimeout(total=15)) as resp:
            if resp.status != 200:
                logger.error("VK API {}: HTTP {}", method, resp.status)
                return None

            data = await resp.json()

            if "error" in data:
                error_info = data["error"]
                logger.error(
                    "VK API {} ошибка {}: {}",
                    method,
                    error_info.get("error_code"),
                    error_info.get("error_msg"),
                )
                return None

            return data.get("response")

    except aiohttp.ClientError as exc:
        logger.error("Сетевая ошибка VK API {}: {}", method, exc)
        return None
    except Exception as exc:
        logger.error("Непредвиденная ошибка VK API {}: {}", method, exc)
        return None


async def _resolve_group_id(
    session: aiohttp.ClientSession,
    group_name: str,
) -> int | None:
    """
    Получить числовой ID группы по короткому имени через utils.resolveScreenName.

    Args:
        group_name: короткое имя группы (без vk.com/)

    Returns:
        Числовой ID группы или None.
    """
    result = await _vk_api_request(
        session,
        "utils.resolveScreenName",
        {"screen_name": group_name},
    )

    if not result:
        # Пустой ответ — группа не найдена
        logger.warning("VK: группа '{}' не найдена через resolveScreenName", group_name)
        return None

    if result.get("type") not in ("group", "page"):
        logger.warning(
            "VK: '{}' — не группа (тип: {})",
            group_name,
            result.get("type"),
        )
        return None

    group_id = result.get("object_id")
    logger.debug("VK: группа '{}' → ID {}", group_name, group_id)
    return group_id


async def fetch_vk_group_posts(
    group_name: str,
    session: aiohttp.ClientSession | None = None,
) -> list[dict[str, Any]]:
    """
    Получить посты вакансий из стены VK-группы.

    Получает последние 50 постов через wall.get, фильтрует по дате
    (последние HOURS_LOOKBACK часов), извлекает зарплаты и ссылки.

    Args:
        group_name: короткое имя группы (например, 'remote_job_ru')
        session: опциональная aiohttp-сессия (если None — создаётся новая)

    Returns:
        Список вакансий в унифицированном формате.
    """
    own_session = session is None
    if own_session:
        session = aiohttp.ClientSession()

    vacancies: list[dict[str, Any]] = []

    try:
        # Получаем ID группы
        group_id = await _resolve_group_id(session, group_name)
        if group_id is None:
            return []

        # Получаем посты со стены
        result = await _vk_api_request(
            session,
            "wall.get",
            {
                "owner_id": f"-{group_id}",
                "count": 50,
                "filter": "owner",
            },
        )

        if not result or "items" not in result:
            logger.warning("VK: пустой ответ wall.get для группы '{}'", group_name)
            return []

        # Порог по дате — только посты за последние HOURS_LOOKBACK часов
        cutoff_ts = (
            datetime.now(tz=timezone.utc) - timedelta(hours=HOURS_LOOKBACK)
        ).timestamp()

        for post in result["items"]:
            # Фильтр по дате
            post_date = post.get("date", 0)
            if post_date < cutoff_ts:
                continue

            text = post.get("text", "")
            if not text or len(text.strip()) < 30:
                continue

            # Заголовок — первая непустая строка
            lines = [ln.strip() for ln in text.split("\n") if ln.strip()]
            title = lines[0] if lines else text[:80]
            if len(title) > 120:
                title = title[:117] + "..."

            # Зарплата
            salary_from, salary_to = _extract_salary(text)

            # Ссылки из текста
            urls = _URL_PATTERN.findall(text)

            # Ссылки из вложений
            for attachment in post.get("attachments", []):
                if attachment.get("type") == "link":
                    link_url = attachment.get("link", {}).get("url", "")
                    if link_url:
                        urls.append(link_url)

            # Прямая ссылка на пост VK
            post_id = post.get("id", 0)
            vk_post_url = f"https://vk.com/wall-{group_id}_{post_id}"

            # Первый найденный URL или ссылка на пост
            url = urls[0] if urls else vk_post_url

            vacancy: dict[str, Any] = {
                "source": "vk",
                "vacancy_id": str(post_id),
                "title": title,
                "company": "",
                "salary_from": salary_from,
                "salary_to": salary_to,
                "description": text,
                "url": url,
                "schedule": "",
                "area": "",
                "published_at": datetime.fromtimestamp(
                    post_date, tz=timezone.utc
                ).isoformat(),
                "extra": {
                    "group": group_name,
                    "vk_post_url": vk_post_url,
                },
            }
            vacancies.append(vacancy)

        logger.info(
            "VK группа '{}': найдено {} записей за последние {} ч",
            group_name,
            len(vacancies),
            HOURS_LOOKBACK,
        )

    except Exception as exc:
        logger.error("Ошибка при парсинге VK-группы '{}': {}", group_name, exc)
    finally:
        if own_session and session:
            await session.close()

    return vacancies


async def fetch_vk_vacancies() -> list[dict[str, Any]]:
    """
    Главная функция модуля — собрать вакансии из всех VK-групп.

    Проверяет наличие VK_ACCESS_TOKEN. Если токен не задан —
    возвращает пустой список с предупреждением.

    Returns:
        Агрегированный список вакансий из всех VK-групп.
    """
    if not VK_ACCESS_TOKEN:
        logger.warning(
            "VK_ACCESS_TOKEN не задан — парсинг VK-групп отключён"
        )
        return []

    if not VK_GROUPS_TO_MONITOR:
        logger.info("Список VK_GROUPS_TO_MONITOR пуст — нечего парсить")
        return []

    all_vacancies: list[dict[str, Any]] = []

    async with aiohttp.ClientSession() as session:
        for group_name in VK_GROUPS_TO_MONITOR:
            try:
                posts = await fetch_vk_group_posts(group_name, session=session)
                all_vacancies.extend(posts)
            except Exception as exc:
                logger.error(
                    "Ошибка при обработке VK-группы '{}': {}",
                    group_name,
                    exc,
                )

    logger.info(
        "Итого из VK собрано {} записей из {} групп",
        len(all_vacancies),
        len(VK_GROUPS_TO_MONITOR),
    )
    return all_vacancies

File: src/sources/vkParser.ts
import { VK } from 'vk-io';
import { config } from '../config/config.js';
import { logger } from '../utils/logger.js';
import { Vacancy } from '../db/database.js';

export async function fetchVkVacancies(): Promise<Vacancy[]> {
  const vacancies: Vacancy[] = [];

  if (!config.VK_ACCESS_TOKEN) {
    logger.debug('VK API: VK_ACCESS_TOKEN is not configured. Skipping VK fetch.');
    return [];
  }

  logger.info('VK API: Starting fetch...');
  
  // Initialize VK SDK
  const vk = new VK({
    token: config.VK_ACCESS_TOKEN,
    apiVersion: config.VK_API_VERSION
  });

  const lookbackHours = config.HOURS_LOOKBACK;
  const cutoffTime = Math.floor((Date.now() - lookbackHours * 60 * 60 * 1000) / 1000); // VK uses Unix timestamps in seconds

  for (const groupName of config.VK_GROUPS_TO_MONITOR) {
    try {
      logger.debug(`VK API: Resolving screen name for group "${groupName}"...`);
      const resolved = await vk.api.utils.resolveScreenName({
        screen_name: groupName
      });

      if (!resolved || resolved.type !== 'group') {
        logger.warn(`VK API: Screen name "${groupName}" resolved to type "${resolved?.type || 'unknown'}", not a group. Skipping.`);
        continue;
      }

      const groupId = resolved.object_id;
      logger.debug(`VK API: Fetching wall posts for group "${groupName}" (ID: -${groupId})...`);
      
      const response = await vk.api.wall.get({
        owner_id: -groupId,
        count: 50,
        filter: 'owner'
      });

      const posts = response.items || [];
      logger.debug(`VK API: Found ${posts.length} posts for "${groupName}".`);

      for (const post of posts) {
        // Skip pinned posts or posts older than cutoffTime
        if (post.is_pinned || !post.date || post.date < cutoffTime) {
          continue;
        }

        const text = post.text || '';
        if (!text.trim()) continue;

        // Extract first line as title
        const lines = text.split('\n').map(l => l.trim()).filter(Boolean);
        const title = lines[0] ? (lines[0].length > 80 ? lines[0].slice(0, 80) + '...' : lines[0]) : 'Вакансия из ВКонтакте';

        // Try to parse salary from text
        let salaryFrom: number | null = null;
        let salaryTo: number | null = null;
        
        // Regex patterns to look for salaries: e.g. 50000, 50 000, 50к, 50k, 50 тыс
        const salaryRangeRegex = /(?:зп|зарплата|оплата|от|до)?\s*(\d+[\s\d]*)\s*(?:-|до)\s*(\d+[\s\d]*)\s*(?:руб|₽|rur|к|k|тыс)/i;
        const rangeMatch = text.match(salaryRangeRegex);

        if (rangeMatch) {
          salaryFrom = parseNumber(rangeMatch[1]);
          salaryTo = parseNumber(rangeMatch[2]);
        } else {
          const singleSalaryRegex = /(?:зп|зарплата|оплата|от|до)?\s*(\d+[\s\d]*)\s*(?:руб|₽|rur|к|k|тыс)/i;
          const singleMatch = text.match(singleSalaryRegex);
          if (singleMatch) {
            const val = parseNumber(singleMatch[1]);
            if (text.toLowerCase().includes('от')) {
              salaryFrom = val;
            } else if (text.toLowerCase().includes('до')) {
              salaryTo = val;
            } else {
              salaryFrom = val;
            }
          }
        }

        // Try to find URLs in attachments or text
        let postUrl = `https://vk.com/wall-${groupId}_${post.id}`;
        let externalUrl = postUrl;

        // Look for link attachments
        const linkAttachment = post.attachments?.find((att: any) => att.type === 'link');
        if (linkAttachment?.link?.url) {
          externalUrl = linkAttachment.link.url;
        } else {
          // Parse link from text via regex
          const urlRegex = /(https?:\/\/[^\s]+)/g;
          const textUrls = text.match(urlRegex);
          if (textUrls && textUrls[0]) {
            externalUrl = textUrls[0];
          }
        }

        vacancies.push({
          source: 'vk',
          vacancy_id: String(post.id),
          title: title,
          company: groupName, // VK group name as fallback company
          salary_from: salaryFrom,
          salary_to: salaryTo,
          url: externalUrl,
          schedule: 'Удалённая работа', // Group parsing is geared towards remote channels
          area: 'РФ',
          description: text,
          requirements: '',
          published_at: new Date(post.date * 1000).toISOString()
        });
      }

    } catch (error: any) {
      logger.error(`VK API Error parsing group "${groupName}": ${error.message}`);
    }
  }

  logger.info(`VK API: Fetched ${vacancies.length} vacancies.`);
  return vacancies;
}

function parseNumber(text: string): number {
  let valStr = text.toLowerCase().replace(/\s/g, '');
  let multiplier = 1;

  if (valStr.endsWith('к') || valStr.endsWith('k')) {
    multiplier = 1000;
    valStr = valStr.slice(0, -1);
  } else if (valStr.endsWith('тыс') || valStr.endsWith('тысяч')) {
    multiplier = 1000;
    valStr = valStr.replace(/(тыс|тысяч)/g, '');
  }

  const num = Number(valStr);
  return isNaN(num) ? 0 : num * multiplier;
}
